### ISSUE TITLE
String::sha256_text() shows up as a hotspot during game launch

### ISSUE THREAD

**Issue by pcbeard** [godotengine/godot#86249]
### Tested versions

Reproducible in:  Godot v4.2.stable - macOS 14.3.0 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (10 Threads)

### System information

Godot v4.2.stable - macOS 14.3.0 - Vulkan (Forward+) - integrated Apple M1 Pro - Apple M1 Pro (10 Threads)

### Issue description

I profiled my app's startup performance and noticed a hot spot in this call sequence:

8	String::utf8() const	
5	String::sha256_text() const	
5	ShaderRD::setup(char const*, char const*, char const*, char const*)	
2	Fsr2AccumulatePassShaderRD::Fsr2AccumulatePassShaderRD()	
2	Fsr2AccumulatePassShaderRD::Fsr2AccumulatePassShaderRD()	
2	RendererRD::FSR2Effect::'unnamed'::()	
2	RendererRD::FSR2Effect::'unnamed'::()	
2	RendererRD::FSR2Effect::FSR2Effect()	
2	RendererRD::FSR2Effect::FSR2Effect()	
2	RendererSceneRenderImplementation::RenderForwardClustered::RenderForwardClustered()	
2	RendererSceneRenderImplementation::RenderForwardClustered::RenderForwardClustered()	
2	RendererCompositorRD::RendererCompositorRD()	
2	RendererCompositorRD::RendererCompositorRD()	
2	RendererCompositorRD::_create_current()	
2	RendererCompositor::create()	
2	RenderingServerDefault::RenderingServerDefault(bool)	
2	RenderingServerDefault::RenderingServerDefault(bool)	
2	Main::setup2()	
2	Main::setup(char const*, int, char**, bool)	

Inspecting `String::sha256_text()` we see it's converting it's character data to utf8 before computing the sha256 value:

```c++
String String::sha256_text() const {
	CharString cs = utf8();
	unsigned char hash[32];
	CryptoCore::sha256((unsigned char *)cs.ptr(), cs.length(), hash);
	return String::hex_encode_buffer(hash, 32);
}
```

Looking at the `.utf8()` method (omitted for brevity), I see that it first scans all the 32-bit code points to calculate the length of the eventual buffer, then it allocates, then it scans again to do the copy. We could do better by changing how the sha256 is computed. Like `String::hash64()` we could compute the sha256 directly using the 32-bit codepoints (or their bytes). Since processors are pretty well adapted to fetching and adding 32-bit values, this would probably end up being a lot faster. Of course, if the algorithm is required to be using the utf8 code points, then we could avoid the copy altogether and write a custom version of sha256 that scans the 32-bit code points directly, converting them to their utf8 versions and computing the hash. Either way would probably be a bit faster than the current approach.

These same optimizations could be made to `String::sha1_text()`, `String::md5_buffer()`, et. al.

### Steps to reproduce

Profiling any Godot app during startup will likely expose these hot spots. Just launch the app while profiling and look for hot spots.

### Minimal reproduction project (MRP)

You'd need a build with profiling enabled to reproduce this. Any trivial Godot project will likely do.

[GodotBouncingBall.zip](https://github.com/godotengine/godot/files/13694799/GodotBouncingBall.zip)

**Comment by pcbeard** [comment]
Looking at the implementation of String itself, which uses `CowData<char32_t>` to hold its elements, I realize that there are 16-bytes of padding preceding every string instance, 8 bytes of which are used to hold 32-bit size and reference counts. We could devote another `uint32_t` to a flags word and record either the utf8 length (e.g. cache the scanned length), or simply store a bit in there that indicates the string is purely ASCII, which would accomplish the same speedup.

**Comment by AThousandShips** [comment]
> I realize that there are 16-bytes of padding preceding every string instance

Where are you seeing this? Are you talking about the data in `CowData`? Because that's not in the stack data so it can't be used in `String`, how do you propose accessing it?

**Comment by pcbeard** [comment]
I tried this change, to avoid rescanning the string if it's already proved to be ASCII-only, and it did reduce the overhead:

```C++
static int _utf8_length(const String &s) {
	int fl = 0, l = s.length();
	const char32_t *d = s.ptr();
	for (int i = 0; i < l; i++) {
		uint32_t c = d[i];
		if (c <= 0x7f) { // 7 bits.
			fl += 1;
		} else if (c <= 0x7ff) { // 11 bits
			fl += 2;
		} else if (c <= 0xffff) { // 16 bits
			fl += 3;
		} else if (c <= 0x001fffff) { // 21 bits
			fl += 4;
		} else if (c <= 0x03ffffff) { // 26 bits
			fl += 5;
			s.print_unicode_error(vformat("Invalid unicode codepoint (%x)", c));
		} else if (c <= 0x7fffffff) { // 31 bits
			fl += 6;
			s.print_unicode_error(vformat("Invalid unicode codepoint (%x)", c));
		} else {
			fl += 1;
			s.print_unicode_error(vformat("Invalid unicode codepoint (%x), cannot represent as UTF-8", c), true);
		}
	}
	return fl;
}

CharString String::utf8() const {
	int l = length();
	if (!l) {
		return CharString();
	}

	int fl = _utf8_length(*this);
	const bool is_ascii = (fl == l);

	CharString utf8s;
	if (fl == 0) {
		return utf8s;
	}

	const char32_t *d = ptr();
	utf8s.resize(fl + 1);
	uint8_t *cdst = (uint8_t *)utf8s.get_data();

	if (is_ascii) {
		while (l--) {
			*cdst++ = (uint8_t)*d++;
		}
		*cdst = '\0';
		return utf8s;
	}

...
}
```

This could probably be accelerated further by SIMD instructions, such as something like this:

```C
void uint32_to_uint8(uint32_t* input, uint8_t* output, size_t length) {
    // Process four uint32_t elements at a time
    for (size_t i = 0; i < length; i += 4) {
        // Load 4 uint32_t elements into a 128-bit register
        __m128i inputVec = _mm_loadu_si128((__m128i*)(input + i));

        // Extract the least significant byte from each 32-bit element
        for (int j = 0; j < 4; ++j) {
            output[i + j] = (uint8_t)_mm_extract_epi8(inputVec, j * 4);
        }
    }
}
```

**Comment by pcbeard** [comment]
@AThousandShips I agree, there's no practical way to manage this flag for mutable strings. I think the best we can do is to either avoid redundant scanning like my change above, or to stop converting to utf8 when computing SHA values. Probably both would be beneficial.

**Comment by AThousandShips** [comment]
Yes but you're ignoring this:
```cpp
		uint64_t *s = (uint64_t *)mem;
		*s = p_bytes;

		uint8_t *s8 = (uint8_t *)mem;

#ifdef DEBUG_ENABLED
		uint64_t new_mem_usage = mem_usage.add(p_bytes);
		max_usage.exchange_if_greater(new_mem_usage);
#endif
		return s8 + PAD_ALIGN;
```
The 8 bytes prior to the data stored by `CowData` is always used, it's not free for anyone's use

**Comment by pcbeard** [comment]
Of course, but am I misunderstanding that PAD_ALIGN is always 16, thus there are 8 more unused bytes? I haven't searched the code beyond cowdata.h to see if there are any other uses.

**Comment by AThousandShips** [comment]
There aren't, there's *always* 8 used for the size of the allocation, see the code above, then there's 8 bytes free for use, which `CowData` uses, the returned pointer is `mem + 16`, and in `mem` you store 8 bytes of `p_bytes`, then at `(mem + 16) - 4` and `(mem + 16) - 8` you store size and ref count respectively, so in total there's `4 + 4 + 8` bytes stored, equaling 16, read the code and see:
https://github.com/godotengine/godot/blob/2d0ee20ff30461b6b10f6fdfba87511a0ebc6642/core/os/memory.cpp#L68-L95

**Comment by AThousandShips** [comment]
I think doing sha256 with implicit conversion from utf32 to utf8, using the data directly, to avoid conversion, this should be trivial because utf32 to utf8 conversion is trivial and simple
